In [1]:
import mysql.connector
import os

# Clean and explore the data

In [128]:

cnx = mysql.connector.connect(
    user='root', 
    password='Mayar123@', 
    host='127.0.0.1', 
    database='md_water_services',
    auth_plugin='mysql_native_password'
)
mycursor = cnx.cursor()

# Execute a query and fetch all results
mycursor.execute('SHOW TABLES')
tables = mycursor.fetchall()
for table in tables:
    print(table)

# Close the cursor and connection
mycursor.close()
cnx.close()


('data_dictionary',)
('employee',)
('global_water_access',)
('location',)
('visits',)
('water_quality',)
('water_source',)
('well_pollution',)


There are 8 tables in md_water_services database

# Updating employee data

We want to add the email addresses to the employee table as first_name.last_name@ndogowater.gov


In [53]:
#return the employee_name values from the employee table with all spaces replaced by periods. For example, 
#if the employee_name is "John Doe", the result will be "John.Doe".
cnx = mysql.connector.connect(
    user='root', 
    password='Mayar123@', 
    host='127.0.0.1', 
    database='md_water_services',
    auth_plugin='mysql_native_password'
)
mycursor = cnx.cursor()

# Execute the query
mycursor.execute("SELECT REPLACE(employee_name, ' ', '.') FROM employee")

# Fetch the results
results = mycursor.fetchall()

# Print the results
for row in results:
    print(row)

# Close the cursor and connection
mycursor.close()
cnx.close()


('Amara.Jengo',)
('Bello.Azibo',)
('Bakari.Iniko',)
('Malachi.Mavuso',)
('Cheche.Buhle',)
('Zuriel.Matembo',)
('Deka.Osumare',)
('Lalitha.Kaburi',)
('Enitan.Zuri',)
('Farai.Nia',)
('Gamba.Shani',)
('Harith.Nyota',)
('Isoke.Amani',)
('Jengo.Tumaini',)
('Kunto.Asha',)
('Lesedi.Kofi',)
('Makena.Thabo',)
('Nia.Furaha',)
('Ona.Sefu',)
('Pili.Zola',)
('Quibilah.Akua',)
('Rudo.Imani',)
('Sanaa.Tendaji',)
('Thandiwe.Kito',)
('Usafi.Ayo',)
('Vuyisile.Kwame',)
('Wambui.Jabali',)
('Xola.Uzuri',)
('Yewande.Ebele',)
('Nadia.Zalika',)
('Ayo.Deka',)
('Bahira.Jamila',)
('Chidi.Kunto',)
('Dalila.Lesedi',)
('Eshe.Makena',)
('Faridah.Khalida',)
('Gavivi.Ona',)
('Hasani.Pili',)
('Iniko.Quibilah',)
('Jengo.Rudo',)
('Kofi.Sanaa',)
('Lulu.Thandiwe',)
('Makena.Usafi',)
('Nia.Vuyisile',)
('Ona.Wambui',)
('Pili.Xola',)
('Quibilah.Yewande',)
('Rudo.Zalika',)
('Sanaa.Ayo',)
('Thabo.Akida',)
('Usafi.Bahir',)
('Vuyisile.Ghadir',)
('Wambui.Hadiya',)
('Xola.Inaya',)
('Yewande.Jengo',)
('Zuri.Kofi',)


In [57]:
#return the employee_name values from the employee table with all spaces replaced by periods and all characters converted to lowercase.
#For example, if the employee_name is "John Doe", the result will be "john.doe".
cnx = mysql.connector.connect(
    user='root', 
    password='Mayar123@', 
    host='127.0.0.1', 
    database='md_water_services',
    auth_plugin='mysql_native_password'
)
mycursor = cnx.cursor()

# Execute the query
mycursor.execute("SELECT LOWER(REPLACE(employee_name, ' ','.')) FROM employee ")

# Fetch the results
results = mycursor.fetchall()

# Print the results
for row in results:
    print(row)

# Close the cursor and connection
mycursor.close()
cnx.close()


('amara.jengo',)
('bello.azibo',)
('bakari.iniko',)
('malachi.mavuso',)
('cheche.buhle',)
('zuriel.matembo',)
('deka.osumare',)
('lalitha.kaburi',)
('enitan.zuri',)
('farai.nia',)
('gamba.shani',)
('harith.nyota',)
('isoke.amani',)
('jengo.tumaini',)
('kunto.asha',)
('lesedi.kofi',)
('makena.thabo',)
('nia.furaha',)
('ona.sefu',)
('pili.zola',)
('quibilah.akua',)
('rudo.imani',)
('sanaa.tendaji',)
('thandiwe.kito',)
('usafi.ayo',)
('vuyisile.kwame',)
('wambui.jabali',)
('xola.uzuri',)
('yewande.ebele',)
('nadia.zalika',)
('ayo.deka',)
('bahira.jamila',)
('chidi.kunto',)
('dalila.lesedi',)
('eshe.makena',)
('faridah.khalida',)
('gavivi.ona',)
('hasani.pili',)
('iniko.quibilah',)
('jengo.rudo',)
('kofi.sanaa',)
('lulu.thandiwe',)
('makena.usafi',)
('nia.vuyisile',)
('ona.wambui',)
('pili.xola',)
('quibilah.yewande',)
('rudo.zalika',)
('sanaa.ayo',)
('thabo.akida',)
('usafi.bahir',)
('vuyisile.ghadir',)
('wambui.hadiya',)
('xola.inaya',)
('yewande.jengo',)
('zuri.kofi',)


In [59]:
#REPLACE: Replacing all spaces (' ') in the employee_name with periods ('.').
#LOWER: Converting all characters in the modified employee_name to lowercase.
#CONCAT: Concatenating the modified employee_name with the string @ndogowater.gov to form a full email address.
A#S new_email: Renaming the resulting column to new_email.
cnx = mysql.connector.connect(
    user='root', 
    password='Mayar123@', 
    host='127.0.0.1', 
    database='md_water_services',
    auth_plugin='mysql_native_password'
)
mycursor = cnx.cursor()

# Execute the query
mycursor.execute("SELECT CONCAT (LOWER(REPLACE(employee_name, ' ', '.')), '@ndogowater.gov') AS new_email  FROM employee")

# Fetch the results
results = mycursor.fetchall()

# Print the results
for row in results:
    print(row)

# Close the cursor and connection
mycursor.close()
cnx.close()


('amara.jengo@ndogowater.gov',)
('bello.azibo@ndogowater.gov',)
('bakari.iniko@ndogowater.gov',)
('malachi.mavuso@ndogowater.gov',)
('cheche.buhle@ndogowater.gov',)
('zuriel.matembo@ndogowater.gov',)
('deka.osumare@ndogowater.gov',)
('lalitha.kaburi@ndogowater.gov',)
('enitan.zuri@ndogowater.gov',)
('farai.nia@ndogowater.gov',)
('gamba.shani@ndogowater.gov',)
('harith.nyota@ndogowater.gov',)
('isoke.amani@ndogowater.gov',)
('jengo.tumaini@ndogowater.gov',)
('kunto.asha@ndogowater.gov',)
('lesedi.kofi@ndogowater.gov',)
('makena.thabo@ndogowater.gov',)
('nia.furaha@ndogowater.gov',)
('ona.sefu@ndogowater.gov',)
('pili.zola@ndogowater.gov',)
('quibilah.akua@ndogowater.gov',)
('rudo.imani@ndogowater.gov',)
('sanaa.tendaji@ndogowater.gov',)
('thandiwe.kito@ndogowater.gov',)
('usafi.ayo@ndogowater.gov',)
('vuyisile.kwame@ndogowater.gov',)
('wambui.jabali@ndogowater.gov',)
('xola.uzuri@ndogowater.gov',)
('yewande.ebele@ndogowater.gov',)
('nadia.zalika@ndogowater.gov',)
('ayo.deka@ndogowater.g

In [137]:
#This command effectively updates each row in the employee table,
#setting the email column to a formatted email address based on the employee_name.

cnx = mysql.connector.connect(
    user='root', 
    password='Mayar123@', 
    host='127.0.0.1', 
    database='md_water_services',
    auth_plugin='mysql_native_password'
)
mycursor = cnx.cursor()

# Execute the update query
mycursor.execute("UPDATE employee SET email = CONCAT(LOWER(REPLACE(employee_name, ' ', '.')), '@ndogowater.gov')")

# Commit the changes
cnx.commit()

# Close the cursor and connection
mycursor.close()
cnx.close()


In [66]:
#The phone numbers should be 12 characters long, consisting of the plus sign, area code (99), and the phone number digits. 
#However, when we use the LENGTH(column) function, it returns 13 characters, indicating there's an extra character 
#That's because there is a space at the end of the number.
cnx = mysql.connector.connect(
    user='root', 
    password='Mayar123@', 
    host='127.0.0.1', 
    database='md_water_services',
    auth_plugin='mysql_native_password'
)
mycursor = cnx.cursor()

# Execute the query
mycursor.execute("SELECT phone_number, LENGTH(phone_number) AS original_length, TRIM(phone_number) AS trimmed_phone_number, LENGTH(TRIM(phone_number)) AS trimmed_length FROM employee;")

# Fetch the results
results = mycursor.fetchall()

# Print the results
for row in results:
    print(row)

# Close the cursor and connection
mycursor.close()
cnx.close()


('+99637993287 ', 13, '+99637993287', 12)
('+99643864786 ', 13, '+99643864786', 12)
('+99222599041 ', 13, '+99222599041', 12)
('+99945849900 ', 13, '+99945849900', 12)
('+99381679640 ', 13, '+99381679640', 12)
('+99034075111 ', 13, '+99034075111', 12)
('+99379364631 ', 13, '+99379364631', 12)
('+99681623240 ', 13, '+99681623240', 12)
('+99248509202 ', 13, '+99248509202', 12)
('+99570082739 ', 13, '+99570082739', 12)
('+99438283153 ', 13, '+99438283153', 12)
('+99917230124 ', 13, '+99917230124', 12)
('+99313603182 ', 13, '+99313603182', 12)
('+99820596465 ', 13, '+99820596465', 12)
('+99176320477 ', 13, '+99176320477', 12)
('+99611183730 ', 13, '+99611183730', 12)
('+99690308755 ', 13, '+99690308755', 12)
('+99522694997 ', 13, '+99522694997', 12)
('+99797494944 ', 13, '+99797494944', 12)
('+99822478933 ', 13, '+99822478933', 12)
('+99533195846 ', 13, '+99533195846', 12)
('+99046972648 ', 13, '+99046972648', 12)
('+99477692836 ', 13, '+99477692836', 12)
('+99895654233 ', 13, '+9989565423

In [68]:

# Establish a connection with the specified auth_plugin
cnx = mysql.connector.connect(
    user='root', 
    password='Mayar123@', 
    host='127.0.0.1', 
    database='md_water_services',
    auth_plugin='mysql_native_password'
)
mycursor = cnx.cursor()

# Execute the query
mycursor.execute("UPDATE employee SET phone_number=TRIM(phone_number)")

# Commit the changes
cnx.commit()

# Close the cursor and connection
mycursor.close()
cnx.close()




The SQL query mycursor.execute("UPDATE employee SET phone_number=TRIM(phone_number)") updates the phone_number column in the employee table by removing any leading or trailing whitespace characters from each phone number entry. 

In [72]:
#View the employees table after cleaning 

# Establish a connection with the specified auth_plugin
cnx = mysql.connector.connect(
    user='root', 
    password='Mayar123@', 
    host='127.0.0.1', 
    database='md_water_services',
    auth_plugin='mysql_native_password'
)
mycursor = cnx.cursor()

# Execute the query
mycursor.execute("SELECT * FROM employee;")

# Fetch the results
results = mycursor.fetchall()

# Print the results
for row in results:
    print(row)

# Close the cursor and connection
mycursor.close()
cnx.close()


(0, 'Amara Jengo', '+99637993287', 'amara.jengo@ndogowater.gov', '36 Pwani Mchangani Road', 'Sokoto', 'Ilanga', 'Field Surveyor')
(1, 'Bello Azibo', '+99643864786', 'bello.azibo@ndogowater.gov', '129 Ziwa La Kioo Road', 'Kilimani', 'Rural', 'Field Surveyor')
(2, 'Bakari Iniko', '+99222599041', 'bakari.iniko@ndogowater.gov', '18 Mlima Tazama Avenue', 'Hawassa', 'Rural', 'Field Surveyor')
(3, 'Malachi Mavuso', '+99945849900', 'malachi.mavuso@ndogowater.gov', '100 Mogadishu Road', 'Akatsi', 'Lusaka', 'Field Surveyor')
(4, 'Cheche Buhle', '+99381679640', 'cheche.buhle@ndogowater.gov', '1 Savanna Street', 'Akatsi', 'Rural', 'Field Surveyor')
(5, 'Zuriel Matembo', '+99034075111', 'zuriel.matembo@ndogowater.gov', '26 Bahari Ya Faraja Road', 'Kilimani', 'Rural', 'Field Surveyor')
(6, 'Deka Osumare', '+99379364631', 'deka.osumare@ndogowater.gov', '104 Kenyatta Street', 'Akatsi', 'Rural', 'Field Surveyor')
(7, 'Lalitha Kaburi', '+99681623240', 'lalitha.kaburi@ndogowater.gov', '145 Sungura Amanpo

# Honouring the workers
# Finding the best

In [140]:
#count how many of our employees live in each town.

cnx = mysql.connector.connect(
    user='root', 
    password='Mayar123@', 
    host='127.0.0.1', 
    database='md_water_services',
    auth_plugin='mysql_native_password'
)
mycursor = cnx.cursor()

# Execute the query to count employees in each town
mycursor.execute("SELECT town_name, COUNT(*) AS employee_count FROM employee GROUP BY town_name")

# Fetch the results
results = mycursor.fetchall()

# Print the results
print("town\tnum_employees")
for row in results:
    print(f"{row[0]}\t{row[1]}")

# Close the cursor and connection
mycursor.close()
cnx.close()


town	num_employees
Ilanga	3
Rural	29
Lusaka	4
Zanzibar	4
Dahabu	6
Kintampo	1
Harare	5
Yaounde	1
Serowe	3


Note how many of our workers are living in smaller communities in the rural parts of MajiNdogo.

In [82]:

cnx = mysql.connector.connect(
    user='root', 
    password='Mayar123@', 
    host='127.0.0.1', 
    database='md_water_services',
    auth_plugin='mysql_native_password'
)
mycursor = cnx.cursor()

# Execute the query to get the top 3 employees with the highest number of visits
mycursor.execute("""
SELECT assigned_employee_id, COUNT(*) AS number_of_visits
FROM visits
WHERE assigned_employee_id IN (0, 1, 2)
GROUP BY assigned_employee_id
ORDER BY number_of_visits DESC
LIMIT 3
""")

# Fetch the results
results = mycursor.fetchall()

# Print the results
print("assigned_employee_id\tnumber_of_visits")
for row in results:
    print(f"{row[0]}\t{row[1]}")

# Close the cursor and connection
mycursor.close()
cnx.close()


assigned_employee_id	number_of_visits
1	3708
2	2033
0	1099


In [90]:
import mysql.connector

# Establish a connection with the specified auth_plugin
cnx = mysql.connector.connect(
    user='root', 
    password='Mayar123@', 
    host='127.0.0.1', 
    database='md_water_services',
    auth_plugin='mysql_native_password'
)
mycursor = cnx.cursor()

# Execute the query to get the top 3 employees with the highest number of visits
mycursor.execute("""
SELECT v.assigned_employee_id,e.assigned_employee_id, COUNT(*) AS number_of_visits, e.employee_name, e.phone_number, e.email
FROM visits AS v
JOIN employee AS e ON v.assigned_employee_id = e.assigned_employee_id
WHERE e.assigned_employee_id IN (0,1,2)
GROUP BY v.assigned_employee_id
ORDER BY number_of_visits DESC
LIMIT 3
""")

# Fetch the results
results = mycursor.fetchall()

# Print the results
print("assigned_employee_id\tnumber_of_visits\temployee_name\t\tphone_number\t\temail")
for row in results:
    print(f"{row[0]}\t\t\t{row[1]}\t\t{row[2]}\t\t{row[3]}\t{row[4]}")

# Close the cursor and connection
mycursor.close()
cnx.close()


assigned_employee_id	number_of_visits	employee_name		phone_number		email
1			1		3708		Bello Azibo	+99643864786
2			2		2033		Bakari Iniko	+99222599041
0			0		1099		Amara Jengo	+99637993287


We notice that Bello Azibo has the highest number of visits

# Analysing Locations

In [92]:
from mysql.connector import Error

# Function to connect to MySQL database
def connect_to_database():
    try:
        # Establish a connection with the specified auth_plugin
        cnx = mysql.connector.connect(
            user='root', 
            password='Mayar123@', 
            host='127.0.0.1', 
            database='md_water_services',
            auth_plugin='mysql_native_password'
        )
        return cnx
    except Error as e:
        print(f"Error connecting to MySQL: {e}")
        return None

# Function to execute SQL query
def execute_query(connection, query):
    cursor = connection.cursor()
    try:
        # Execute the query
        cursor.execute(query)
        # Fetch all the results
        results = cursor.fetchall()
        return results
    except Error as e:
        print(f"Error executing query: {e}")
        return None
    finally:
        cursor.close()

# Connect to the database
connection = connect_to_database()

if connection:
    try:
        # SQL query to count records per town and order by count descending
        query = """
        SELECT town_name, COUNT(*) AS records_per_town
        FROM location
        GROUP BY town_name
        ORDER BY records_per_town DESC;
        """
        
        # Execute the query
        results = execute_query(connection, query)
        
        # Print the results
        if results:
            print("town_name\trecords_per_town")
            for row in results:
                print(f"{row[0]}\t\t{row[1]}")
    finally:
        # Close the database connection
        connection.close()
else:
    print("Unable to connect to the database.")


town_name	records_per_town
Rural		23740
Harare		1650
Amina		1090
Lusaka		1070
Mrembo		990
Asmara		930
Dahabu		930
Kintampo		780
Ilanga		780
Isiqalo		770
Zuri		750
Amara		710
Majengo		620
Pwani		520
Kofi		510
Serowe		480
Bahari		470
Marang		460
Abidjan		400
Bello		400
Cheche		370
Deka		350
Djenne		310
Zanzibar		300
Yaounde		270


In [93]:


# Function to connect to MySQL database
def connect_to_database():
    try:
        # Establish a connection with the specified auth_plugin
        cnx = mysql.connector.connect(
            user='root', 
            password='Mayar123@', 
            host='127.0.0.1', 
            database='md_water_services',
            auth_plugin='mysql_native_password'
        )
        return cnx
    except Error as e:
        print(f"Error connecting to MySQL: {e}")
        return None

# Function to execute SQL query
def execute_query(connection, query):
    cursor = connection.cursor()
    try:
        # Execute the query
        cursor.execute(query)
        # Fetch all the results
        results = cursor.fetchall()
        return results
    except Error as e:
        print(f"Error executing query: {e}")
        return None
    finally:
        cursor.close()

# Connect to the database
connection = connect_to_database()

if connection:
    try:
        # SQL query to count records per town and order by count descending
        query = """
        SELECT province_name, COUNT(*) AS records_per_town
        FROM location
        GROUP BY province_name
        ORDER BY records_per_town DESC;
        """
        
        # Execute the query
        results = execute_query(connection, query)
        
        # Print the results
        if results:
            print("town_name\trecords_per_town")
            for row in results:
                print(f"{row[0]}\t\t{row[1]}")
    finally:
        # Close the database connection
        connection.close()
else:
    print("Unable to connect to the database.")


town_name	records_per_town
Kilimani		9510
Akatsi		8940
Sokoto		8220
Amanzi		6950
Hawassa		6030



From this table, it's pretty clear that most of the water sources in the survey are situated in small rural communities, scattered across Maji Ndogo.

In [94]:


# Function to connect to MySQL database
def connect_to_database():
    try:
        # Establish a connection with the specified auth_plugin
        cnx = mysql.connector.connect(
            user='root', 
            password='Mayar123@', 
            host='127.0.0.1', 
            database='md_water_services',
            auth_plugin='mysql_native_password'
        )
        return cnx
    except Error as e:
        print(f"Error connecting to MySQL: {e}")
        return None

# Function to execute SQL query
def execute_query(connection, query):
    cursor = connection.cursor()
    try:
        # Execute the query
        cursor.execute(query)
        # Fetch all the results
        results = cursor.fetchall()
        return results
    except Error as e:
        print(f"Error executing query: {e}")
        return None
    finally:
        cursor.close()

# Connect to the database
connection = connect_to_database()

if connection:
    try:
        # SQL query to retrieve province_name, town_name, and count of records per town
        query = """
        SELECT province_name, town_name, COUNT(*) AS records_per_town
        FROM location
        GROUP BY province_name, town_name
        ORDER BY province_name ASC, records_per_town DESC;
        """
        
        # Execute the query
        results = execute_query(connection, query)
        
        # Print the results
        if results:
            print("province_name\ttown_name\trecords_per_town")
            for row in results:
                print(f"{row[0]}\t\t{row[1]}\t\t{row[2]}")
    finally:
        # Close the database connection
        connection.close()
else:
    print("Unable to connect to the database.")


province_name	town_name	records_per_town
Akatsi		Rural		6290
Akatsi		Lusaka		1070
Akatsi		Harare		800
Akatsi		Kintampo		780
Amanzi		Rural		3100
Amanzi		Asmara		930
Amanzi		Dahabu		930
Amanzi		Amina		670
Amanzi		Pwani		520
Amanzi		Abidjan		400
Amanzi		Bello		400
Hawassa		Rural		3900
Hawassa		Serowe		480
Hawassa		Amina		420
Hawassa		Deka		350
Hawassa		Djenne		310
Hawassa		Zanzibar		300
Hawassa		Yaounde		270
Kilimani		Rural		5440
Kilimani		Mrembo		990
Kilimani		Harare		850
Kilimani		Isiqalo		770
Kilimani		Zuri		750
Kilimani		Amara		710
Sokoto		Rural		5010
Sokoto		Ilanga		780
Sokoto		Majengo		620
Sokoto		Kofi		510
Sokoto		Bahari		470
Sokoto		Marang		460
Sokoto		Cheche		370


These results show us that the field surveyors did an excellent job of documenting the status of our country's water crisis. Every province and town has many documented sources.

In [96]:
#seeing number of records for each location tyoe

# Function to connect to MySQL database
def connect_to_database():
    try:
        # Establish a connection with the specified auth_plugin
        cnx = mysql.connector.connect(
            user='root', 
            password='Mayar123@', 
            host='127.0.0.1', 
            database='md_water_services',
            auth_plugin='mysql_native_password'
        )
        return cnx
    except Error as e:
        print(f"Error connecting to MySQL: {e}")
        return None

# Function to execute SQL query
def execute_query(connection, query):
    cursor = connection.cursor()
    try:
        # Execute the query
        cursor.execute(query)
        # Fetch all the results
        results = cursor.fetchall()
        return results
    except Error as e:
        print(f"Error executing query: {e}")
        return None
    finally:
        cursor.close()

# Connect to the database
connection = connect_to_database()

if connection:
    try:
        query = """
        SELECT location_type, COUNT(*) AS records_per_town
        FROM location
        GROUP BY location_type
        """
        
        # Execute the query
        results = execute_query(connection, query)
        
        # Print the results
        if results:
            print("num_of_sources\tlocation_type")
            for row in results:
                print(f"{row[0]}\t\t{row[1]}")
    finally:
        # Close the database connection
        connection.close()
else:
    print("Unable to connect to the database.")


num_of_sources	location_type
Urban		15910
Rural		23740


We can see that there are more rural sources than urban, but it's really hard to understand those numbers. Percentages are more relatable. If we use SQL as a very overpowered calculator:
SELECT 23740 / (15910 + 23740) * 100
We can see that 60% of all water sources in the data set are in rural communities.

 what are some of the insights we gained from the location table?
1. Our entire country was properly canvassed, and our dataset represents the situation on the ground.
2. 60% of our water sources are in rural communities across Maji Ndogo. We need to keep this in mind when we make decisions.

# Diving into the sources

In [98]:
#how many people were surveyed in total

# Function to connect to MySQL database
def connect_to_database():
    try:
        # Establish a connection with the specified auth_plugin
        cnx = mysql.connector.connect(
            user='root', 
            password='Mayar123@', 
            host='127.0.0.1', 
            database='md_water_services',
            auth_plugin='mysql_native_password'
        )
        return cnx
    except Error as e:
        print(f"Error connecting to MySQL: {e}")
        return None

# Function to execute SQL query
def execute_query(connection, query):
    cursor = connection.cursor()
    try:
        # Execute the query
        cursor.execute(query)
        # Fetch all the results
        results = cursor.fetchall()
        return results
    except Error as e:
        print(f"Error executing query: {e}")
        return None
    finally:
        cursor.close()

# Connect to the database
connection = connect_to_database()

if connection:
    try:
        # SQL query to retrieve total number of people surveyed from water_source table
        query = """
       SELECT SUM(number_of_people_served) AS total_people_surveyed
       FROM water_source;
        """
        
        # Execute the query
        results = execute_query(connection, query)
        
        # Print the results
        if results:
            print("total_people_surveyed")
            for row in results:
                print(f"{row[0]}")
    finally:
        # Close the database connection
        connection.close()
else:
    print("Unable to connect to the database.")


total_people_surveyed
27628140


In [104]:
#count how many of each of the different water source types there are
import mysql.connector
from mysql.connector import Error

# Function to connect to MySQL database
def connect_to_database():
    try:
        # Establish a connection with the specified auth_plugin
        cnx = mysql.connector.connect(
            user='root', 
            password='Mayar123@', 
            host='127.0.0.1', 
            database='md_water_services',
            auth_plugin='mysql_native_password'
        )
        return cnx
    except Error as e:
        print(f"Error connecting to MySQL: {e}")
        return None

# Function to execute SQL query
def execute_query(connection, query):
    cursor = connection.cursor()
    try:
        # Execute the query
        cursor.execute(query)
        # Fetch all the results
        results = cursor.fetchall()
        return results
    except Error as e:
        print(f"Error executing query: {e}")
        return None
    finally:
        cursor.close()

# Connect to the database
connection = connect_to_database()

if connection:
    try:
        # SQL query to retrieve province_name, town_name, and count of records per town
        query = """
       SELECT type_of_water_source, COUNT(*) AS count_per_type
FROM water_source
GROUP BY type_of_water_source
ORDER BY count_per_type DESC;
        """
        
        # Execute the query
        results = execute_query(connection, query)
        
        # Print the results
        if results:
            print("Water Source Types\tCount")
            for row in results:
                print(f"{row[0]}\t\t\t{row[1]}")
        
        
    finally:
        # Close the database connection
        connection.close()
else:
    print("Unable to connect to the database.")


Water Source Types	Count
well			17383
tap_in_home			7265
tap_in_home_broken			5856
shared_tap			5767
river			3379


Which of those sources stands out? It is pretty clear that although there was a drought, water is still abundant in Maji Ndogo. This isn't just an informative result, we will need these numbers to understand how much all of these repairs will cost. If we know how many taps we need to install, and we know how much it will cost to install them, we can calculate how much it will cost to solve the water crisis.

In [105]:
#What is the average number of people that are served by each water source?

import mysql.connector
from mysql.connector import Error

# Function to connect to MySQL database
def connect_to_database():
    try:
        # Establish a connection with the specified auth_plugin
        cnx = mysql.connector.connect(
            user='root', 
            password='Mayar123@', 
            host='127.0.0.1', 
            database='md_water_services',
            auth_plugin='mysql_native_password'
        )
        return cnx
    except Error as e:
        print(f"Error connecting to MySQL: {e}")
        return None

# Function to execute SQL query
def execute_query(connection, query):
    cursor = connection.cursor()
    try:
        # Execute the query
        cursor.execute(query)
        # Fetch all the results
        results = cursor.fetchall()
        return results
    except Error as e:
        print(f"Error executing query: {e}")
        return None
    finally:
        cursor.close()

# Connect to the database
connection = connect_to_database()

if connection:
    try:
        #This query provides insights into which types of water sources are serving the most people on average
        query = """
     SELECT type_of_water_source, 
       ROUND(SUM(number_of_people_served) / COUNT(*)) AS average_people_served
FROM water_source
GROUP BY type_of_water_source
ORDER BY average_people_served DESC;
        """
        results = execute_query(connection, query)
        
        # Print the results
        if results:
            print("type_of_water_source\tAverage_people_Served")
            for row in results:
                print(f"{row[0]}\t\t\t{row[1]}")
        
        
    finally:
        # Close the database connection
        connection.close()
else:
    print("Unable to connect to the database.")


type_of_water_source	Average_people_Served
shared_tap			2071
river			699
tap_in_home_broken			649
tap_in_home			644
well			279


These results are telling us that 644 people share a tap_in_home onaverage. Doesthatmakesense?
No it doesn’t, right?
Remember I told you a few important things that apply to tap_in_home and broken_tap_in_home? The surveyors combined the data of many households together and added this as a single tap record, but each household actually has its own tap. In addition to this, there is an average of 6 people living in a home. So 6 people actually share 1 tap (not 644).

This means that 1 tap_in_home actually represents 644 ÷ 6 = ± 100 taps.



On the other hand, 2000 share a single public tap on average! So looking at these results, we probably should focus on improving shared taps first.

In [106]:
#the total number of people served by each type of water source in total,

# Function to connect to MySQL database
def connect_to_database():
    try:
        # Establish a connection with the specified auth_plugin
        cnx = mysql.connector.connect(
            user='root', 
            password='Mayar123@', 
            host='127.0.0.1', 
            database='md_water_services',
            auth_plugin='mysql_native_password'
        )
        return cnx
    except Error as e:
        print(f"Error connecting to MySQL: {e}")
        return None

# Function to execute SQL query
def execute_query(connection, query):
    cursor = connection.cursor()
    try:
        # Execute the query
        cursor.execute(query)
        # Fetch all the results
        results = cursor.fetchall()
        return results
    except Error as e:
        print(f"Error executing query: {e}")
        return None
    finally:
        cursor.close()

# Connect to the database
connection = connect_to_database()

if connection:
    try:
        # Task: Calculate the total number of people served by each type of water source
        query_task = """
        SELECT type_of_water_source, 
               SUM(number_of_people_served) AS total_people_served
        FROM water_source
        GROUP BY type_of_water_source
        ORDER BY total_people_served DESC;
        """
        
        # Execute the query
        results_task = execute_query(connection, query_task)
        
        # Print the results
        if results_task:
            print("Total People Served by Each Water Source Type")
            print("Water Source Types\tTotal People Served")
            for row in results_task:
                print(f"{row[0]}\t\t\t{row[1]}")
        else:
            print("No data found for water source types.")
            
    finally:
        # Close the database connection
        connection.close()
else:
    print("Unable to connect to the database.")


Total People Served by Each Water Source Type
Water Source Types	Total People Served
shared_tap			11945272
well			4841724
tap_in_home			4678880
tap_in_home_broken			3799720
river			2362544


In [114]:

#Calculate percentages of people served by each water source type

# Function to connect to MySQL database
def connect_to_database():
    try:
        # Establish a connection with the specified auth_plugin
        cnx = mysql.connector.connect(
            user='root', 
            password='Mayar123@', 
            host='127.0.0.1', 
            database='md_water_services',
            auth_plugin='mysql_native_password'
        )
        return cnx
    except Error as e:
        print(f"Error connecting to MySQL: {e}")
        return None

# Function to execute SQL query
def execute_query(connection, query):
    cursor = connection.cursor()
    try:
        # Execute the query
        cursor.execute(query)
        # Fetch all the results
        results = cursor.fetchall()
        return results
    except Error as e:
        print(f"Error executing query: {e}")
        return None
    finally:
        cursor.close()

# Connect to the database
connection = connect_to_database()

if connection:
    try:
        # SQL query to calculate percentages of people served by each water source type
        query = """
        SELECT 
            type_of_water_source, 
            SUM(number_of_people_served) AS total_people_served,
            ROUND((SUM(number_of_people_served) / (
                SELECT SUM(number_of_people_served) FROM water_source
            ) * 100),0) AS percentage_served
        FROM water_source
        GROUP BY type_of_water_source
        ORDER BY total_people_served DESC;
        """
        
        # Execute the query
        results = execute_query(connection, query)
        
        # Print the results
        if results:
            print("Water Source Types\tTotal People Served\tPercentage Served")
            for row in results:
                print(f"{row[0]}\t\t\t{row[1]}\t\t\t{row[2]:.2f}%")
        else:
            print("No data found for water source types.")
            
    finally:
        # Close the database connection
        connection.close()
else:
    print("Unable to connect to the database.")


Water Source Types	Total People Served	Percentage Served
shared_tap			11945272			43.00%
well			4841724			18.00%
tap_in_home			4678880			17.00%
tap_in_home_broken			3799720			14.00%
river			2362544			9.00%



43% of our people are using shared taps in their communities,and on average,we saw earlier,that 2000 people share oneshared_tap.

By adding tap_in_home and tap_in_home_broken together, we see that 31% of people have water infrastructure installed in their homes, but 45% (14/31) of these taps are not working! This isn't the tap itself that is broken, but rather the infrastructure like treatment plants, reservoirs, pipes, and pumps that serve these homes that are broken.
18% of people are using wells. But only 4916 out of 17383 are clean = 28% 

# Thinking about how we can repair

At some point, we will have to fix or improve all of the infrastructure, so we should start thinking about how we can make a data-driven decision how to do it. I think a simple approach is to fix the things that affect most people first. So let's write a query that ranks each type of source based on how many people in total use it. RANK() should tell you we are going to need a window function to do this, so let's think through the problem.

In [117]:


# Function to connect to MySQL database
def connect_to_database():
    try:
        # Establish a connection with the specified auth_plugin
        cnx = mysql.connector.connect(
            user='root', 
            password='Mayar123@', 
            host='127.0.0.1', 
            database='md_water_services',
            auth_plugin='mysql_native_password'
        )
        return cnx
    except Error as e:
        print(f"Error connecting to MySQL: {e}")
        return None

# Function to execute SQL query
def execute_query(connection, query):
    cursor = connection.cursor()
    try:
        # Execute the query
        cursor.execute(query)
        # Fetch all the results
        results = cursor.fetchall()
        return results
    except Error as e:
        print(f"Error executing query: {e}")
        return None
    finally:
        cursor.close()

# Connect to the database
connection = connect_to_database()

if connection:
    try:
        # Set the rank variable and execute the query to rank water sources by the total people served, excluding 'tap_in_home'
        queries = [
            "SET @rank = 0;",
            """
            SELECT 
                type_of_water_source,
                total_people_served,
                @rank := @rank + 1 AS rank_by_population
            FROM (
                SELECT 
                    type_of_water_source,
                    SUM(number_of_people_served) AS total_people_served,
                    ROUND((SUM(number_of_people_served) / (
                        SELECT SUM(number_of_people_served) FROM water_source
                    ) * 100), 0) AS percentage_served
                FROM water_source
                WHERE type_of_water_source != 'tap_in_home'
                GROUP BY type_of_water_source
                ORDER BY total_people_served DESC
            ) AS subquery;
            """
        ]
        
        # Execute each query
        for query in queries:
            results = execute_query(connection, query)
        
        # Print the results
        if results:
            print("Water Source Types\tTotal People Served\tRank")
            for row in results:
                print(f"{row[0]}\t\t\t{row[1]}\t\t\t{row[2]}")
        else:
            print("No data found for water source types.")
            
    finally:
        # Close the database connection
        connection.close()
else:
    print("Unable to connect to the database.")


Error executing query: No result set to fetch from.
Water Source Types	Total People Served	Rank
shared_tap			11945272			1
well			4841724			2
tap_in_home_broken			3799720			3
river			2362544			4


If someone has a tap in their home, they already have the best source available. Since we can’t do anything more to improve this, we should remove tap_in_home from the ranking before we continue.

Ok, so we should fix shared taps first, then wells, and so on. But the next question is, which shared taps or wells should be fixed first? We can use the same logic; the most used sources should really be fixed first.


In [118]:


# Function to connect to MySQL database
def connect_to_database():
    try:
        # Establish a connection with the specified auth_plugin
        cnx = mysql.connector.connect(
            user='root', 
            password='Mayar123@', 
            host='127.0.0.1', 
            database='md_water_services',
            auth_plugin='mysql_native_password'
        )
        return cnx
    except Error as e:
        print(f"Error connecting to MySQL: {e}")
        return None

# Function to execute SQL query
def execute_query(connection, query):
    cursor = connection.cursor()
    try:
        # Execute the query
        cursor.execute(query)
        # Fetch all the results
        results = cursor.fetchall()
        return results
    except Error as e:
        print(f"Error executing query: {e}")
        return None
    finally:
        cursor.close()

# Connect to the database
connection = connect_to_database()

if connection:
    try:
        # SQL query to rank water sources by the total people served, excluding 'tap_in_home'
        query = """
        SELECT 
            source_id,
            type_of_water_source,
            number_of_people_served,
            RANK() OVER (PARTITION BY type_of_water_source ORDER BY number_of_people_served DESC) AS priority_rank
        FROM water_source
        WHERE type_of_water_source != 'tap_in_home'
        ORDER BY priority_rank, number_of_people_served DESC;
        """
        
        # Execute the query
        results = execute_query(connection, query)
        
        # Print the results
        if results:
            print("source_id\ttype_of_water_source\tnumber_of_people_served\tpriority_rank")
            for row in results:
                print(f"{row[0]}\t{row[1]}\t{row[2]}\t{row[3]}")
        else:
            print("No data found for water source types.")
            
    finally:
        # Close the database connection
        connection.close()
else:
    print("Unable to connect to the database.")


source_id	type_of_water_source	number_of_people_served	priority_rank
AkRu05603224	shared_tap	3998	1
HaRu19509224	shared_tap	3998	1
SoRu36238224	river	998	1
SoRu36791224	river	998	1
SoRu36880224	river	998	1
SoRu34798224	river	998	1
SoRu35837224	river	998	1
SoRu37756224	river	998	1
SoRu38142224	river	998	1
KiRu30353224	river	998	1
SoIl32972224	river	998	1
SoMa33775224	river	998	1
KiZu31236224	river	998	1
KiRu28591224	river	998	1
KiMr25030224	river	998	1
SoRu38890224	tap_in_home_broken	998	1
KiHa22638224	tap_in_home_broken	998	1
KiIs23844224	tap_in_home_broken	998	1
KiMr24493224	tap_in_home_broken	998	1
AmRu15500224	tap_in_home_broken	998	1
SoMa33958224	tap_in_home_broken	998	1
SoIl32412224	tap_in_home_broken	998	1
HaRu20646224	tap_in_home_broken	998	1
HaRu19976224	tap_in_home_broken	998	1
AmPw12633224	tap_in_home_broken	998	1
AmAb09192224	tap_in_home_broken	998	1
AmAm09926224	tap_in_home_broken	998	1
AmAm09750224	tap_in_home_broken	998	1
SoRu36513224	tap_in_home_broken	998	1
KiRu29234224

# Analysing queues

The visits table documented all of the visits our field surveyors made to each location. For most sources, one visit was enough, but if there were queues,they visited the location a couple of times to get a good idea of the time it took for people to queue for water. So we have the time that they collected the data, how many times the site was visited, and how long people had to queue for water.
So, we look at the information we have available, and think of what we could learn from it.

these are some of the things worth looking at: 
1. How long did the survey take?
2. What is the average total queue time for water?
3. What is the average queue time on different days?
4. How can we communicate this information efficiently?

In [120]:
#get the time of the survey

# Function to connect to MySQL database
def connect_to_database():
    try:
        # Establish a connection with the specified auth_plugin
        cnx = mysql.connector.connect(
            user='root', 
            password='Mayar123@', 
            host='127.0.0.1', 
            database='md_water_services',
            auth_plugin='mysql_native_password'
        )
        return cnx
    except Error as e:
        print(f"Error connecting to MySQL: {e}")
        return None

# Function to execute SQL query
def execute_query(connection, query):
    cursor = connection.cursor()
    try:
        # Execute the query
        cursor.execute(query)
        # Fetch all the results
        results = cursor.fetchall()
        return results
    except Error as e:
        print(f"Error executing query: {e}")
        return None
    finally:
        cursor.close()

# Connect to the database
connection = connect_to_database()

if connection:
    try:
        # SQL query to calculate the survey duration
        query = """
        SELECT 
            MIN(time_of_record) AS start_date, 
            MAX(time_of_record) AS end_date, 
            DATEDIFF(MAX(time_of_record), MIN(time_of_record)) AS survey_duration
        FROM visits;
        """
        
        # Execute the query
        results = execute_query(connection, query)
        
        # Print the results
        if results:
            print("start_date\tend_date\tsurvey_duration")
            for row in results:
                print(f"{row[0]}\t{row[1]}\t{row[2]} days")
        else:
            print("No data found for survey dates.")
            
    finally:
        # Close the database connection
        connection.close()
else:
    print("Unable to connect to the database.")


start_date	end_date	survey_duration
2021-01-01 09:10:00	2023-07-14 13:53:00	924 days


I get 924 days which is about 2 and a half years!

In [2]:
#how long people have to queue on average in Maji Ndogo.

# Function to connect to MySQL database
def connect_to_database():
    try:
        # Establish a connection with the specified auth_plugin
        cnx = mysql.connector.connect(
            user='root', 
            password='Mayar123@', 
            host='127.0.0.1', 
            database='md_water_services',
            auth_plugin='mysql_native_password'
        )
        return cnx
    except Error as e:
        print(f"Error connecting to MySQL: {e}")
        return None

# Function to execute SQL query
def execute_query(connection, query):
    cursor = connection.cursor()
    try:
        # Execute the query
        cursor.execute(query)
        # Fetch all the results
        results = cursor.fetchall()
        return results
    except Error as e:
        print(f"Error executing query: {e}")
        return None
    finally:
        cursor.close()

# Connect to the database
connection = connect_to_database()

if connection:
    try:
        # SQL query to calculate the average queue time in Maji Ndogo
        query = """
        SELECT 
            AVG(NULLIF(time_in_queue, 0)) AS average_queue_time
        FROM visits;
        """
        
        # Execute the query
        results = execute_query(connection, query)
        
        # Print the results
        if results:
            print("average_queue_time")
            for row in results:
                print(f"{row[0]:.2f} minutes")
        else:
            print("No data found for queue times.")
            
    finally:
        # Close the database connection
        connection.close()
else:
    print("Unable to connect to the database.")


average_queue_time
123.26 minutes


You should get a queue time of about  123 min. So on average,people take two hours to fetch water if they don't have a tap in their homes.

That may sound reasonable, but some days might have more people who need water, and only have time to go and collect some on certain days.

In [121]:
# queue times aggregated across the different days of the week.

In [124]:


# Function to connect to MySQL database
def connect_to_database():
    try:
        # Establish a connection with the specified auth_plugin
        cnx = mysql.connector.connect(
            user='root', 
            password='Mayar123@', 
            host='127.0.0.1', 
            database='md_water_services',
            auth_plugin='mysql_native_password'
        )
        return cnx
    except Error as e:
        print(f"Error connecting to MySQL: {e}")
        return None

# Function to execute SQL query
def execute_query(connection, query):
    cursor = connection.cursor()
    try:
        # Execute the query
        cursor.execute(query)
        # Fetch all the results
        results = cursor.fetchall()
        return results
    except Error as e:
        print(f"Error executing query: {e}")
        return None
    finally:
        cursor.close()

# Connect to the database
connection = connect_to_database()

if connection:
    try:
        # SQL query to calculate the average queue time by day of the week
        query = """
        SELECT 
            DAYNAME(time_of_record) AS day_of_week,
            ROUND(AVG(NULLIF(time_in_queue, 0))) AS average_queue_time
        FROM visits
        GROUP BY day_of_week
        ORDER BY FIELD(day_of_week, 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday');
        """
        
        # Execute the query
        results = execute_query(connection, query)
        
        # Print the results
        if results:
            print("day_of_week\taverage_queue_time")
            for row in results:
                print(f"{row[0]}\t{row[1]:.2f} minutes")
        else:
            print("No data found for queue times.")
            
    finally:
        # Close the database connection
        connection.close()
else:
    print("Unable to connect to the database.")


day_of_week	average_queue_time
Monday	137.00 minutes
Tuesday	108.00 minutes
Wednesday	97.00 minutes
Thursday	105.00 minutes
Friday	120.00 minutes
Saturday	246.00 minutes
Sunday	82.00 minutes


Saturdays have much longer queue times compared to the other days

In [125]:
#We can also look at what time during the day people collect water.

# Function to connect to MySQL database
def connect_to_database():
    try:
        # Establish a connection with the specified auth_plugin
        cnx = mysql.connector.connect(
            user='root', 
            password='Mayar123@', 
            host='127.0.0.1', 
            database='md_water_services',
            auth_plugin='mysql_native_password'
        )
        return cnx
    except Error as e:
        print(f"Error connecting to MySQL: {e}")
        return None

# Function to execute SQL query
def execute_query(connection, query):
    cursor = connection.cursor()
    try:
        # Execute the query
        cursor.execute(query)
        # Fetch all the results
        results = cursor.fetchall()
        return results
    except Error as e:
        print(f"Error executing query: {e}")
        return None
    finally:
        cursor.close()

# Connect to the database
connection = connect_to_database()

if connection:
    try:
        # SQL query to calculate the average queue time by hour of the day
        query = """
       SELECT 
    HOUR(time_of_record) AS hour_of_day,
    ROUND(AVG(NULLIF(time_in_queue, 0))) AS average_queue_time
FROM visits
GROUP BY hour_of_day
ORDER BY hour_of_day;
        """
        
        # Execute the query
        results = execute_query(connection, query)
        
        # Print the results
        if results:
            print("hour_of_day\taverage_queue_time")
            for row in results:
                print(f"{row[0]}\t{row[1]:.2f} minutes")
        else:
            print("No data found for queue times.")
            
    finally:
        # Close the database connection
        connection.close()
else:
    print("Unable to connect to the database.")


hour_of_day	average_queue_time
6	149.00 minutes
7	149.00 minutes
8	149.00 minutes
9	118.00 minutes
10	114.00 minutes
11	111.00 minutes
12	112.00 minutes
13	115.00 minutes
14	114.00 minutes
15	114.00 minutes
16	114.00 minutes
17	149.00 minutes
18	147.00 minutes
19	168.00 minutes


mornings and evenings are the busiest. It looks like people collect water before and after work. Wouldn't it be nice to break down the queue times for each hour of each day? In a spreadsheet, we can just create a pivot table.


In [127]:
#breaking  down the queue times for each hour of each day

# Function to connect to MySQL database
def connect_to_database():
    try:
        # Establish a connection with the specified auth_plugin
        cnx = mysql.connector.connect(
            user='root', 
            password='Mayar123@', 
            host='127.0.0.1', 
            database='md_water_services',
            auth_plugin='mysql_native_password'
        )
        return cnx
    except Error as e:
        print(f"Error connecting to MySQL: {e}")
        return None

# Function to execute SQL query
def execute_query(connection, query):
    cursor = connection.cursor()
    try:
        # Execute the query
        cursor.execute(query)
        # Fetch all the results
        results = cursor.fetchall()
        # Get the column names
        column_names = [desc[0] for desc in cursor.description]
        return results, column_names
    except Error as e:
        print(f"Error executing query: {e}")
        return None, None
    finally:
        cursor.close()

# SQL query to retrieve average queue times by hour and day of the week
query = """
SELECT
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Sunday' THEN time_in_queue ELSE NULL END), 0) AS Sunday,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Monday' THEN time_in_queue ELSE NULL END), 0) AS Monday,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Tuesday' THEN time_in_queue ELSE NULL END), 0) AS Tuesday,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Wednesday' THEN time_in_queue ELSE NULL END), 0) AS Wednesday,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Thursday' THEN time_in_queue ELSE NULL END), 0) AS Thursday,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Friday' THEN time_in_queue ELSE NULL END), 0) AS Friday,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Saturday' THEN time_in_queue ELSE NULL END), 0) AS Saturday
FROM
    visits
WHERE
    time_in_queue != 0 -- this excludes other sources with 0 queue times
GROUP BY
    hour_of_day
ORDER BY
    hour_of_day;
"""

# Connect to the database
connection = connect_to_database()

if connection:
    try:
        # Execute the query
        results, column_names = execute_query(connection, query)
        
        # Print the results
        if results and column_names:
            print("\t".join(column_names))
            for row in results:
                print("\t".join(str(cell) for cell in row))
    finally:
        # Close the database connection
        connection.close()
else:
    print("Unable to connect to the database.")


hour_of_day	Sunday	Monday	Tuesday	Wednesday	Thursday	Friday	Saturday
06:00	79	190	134	112	134	153	247
07:00	82	186	128	111	139	156	247
08:00	86	183	130	119	129	153	247
09:00	84	127	105	94	99	107	252
10:00	83	119	99	89	95	112	259
11:00	78	115	102	86	99	104	236
12:00	78	115	97	88	96	109	239
13:00	81	122	97	98	101	115	242
14:00	83	127	104	92	96	110	244
15:00	83	126	104	88	92	110	248
16:00	83	127	99	90	99	109	251
17:00	79	181	135	121	129	151	251
18:00	80	174	122	113	132	158	240
19:00	127	159	145	176	137	103	282


1. Queues are very long on a Monday morning and Monday evening as people rush to get water.
2. Wednesdayhasthelowestqueuetimes,butlongqueuesonWednesdayevening.
3. People have to queue pretty much twice as long on Saturdays compared to the weekdays. It looks like people spend their Saturdays queueing for water, perhaps for the week's supply?
4. The shortest queues are on Sundays, and this is a cultural thing. The people of Maji Ndogo prioritise family and religion, so Sundays are spent with family and friends.

# Insights


1. Most water sources are rural.
2. 43% of our people are using shared taps. 2000 people often share one tap.
3. 31% of our population has water infrastructure in their homes, but within that group, 45% face non-functional systems due to issues with pipes, pumps, and reservoirs.
4. 18% of our people are using wells of which, but within that, only 28% are clean..
5. Our citizens often face long wait times for water, averaging more than 120 minutes.
6. In terms of queues:
- Queues are very long on Saturdays.
-Queuesarelongerinthemorningsandevenings.
- Wednesdays and Sundays have the shortest queues.

# Start of the plan


We have started thinking about a plan:
1. We want to focus our efforts on improving the water sources that affect the most people.
- Most people will benefit if we improve the shared taps first.
- Wells are a good source of water, but many are contaminated. Fixing this will benefit a lot of people.
- Fixing existing infrastructure will help many people. If they have running water again, they won't have to queue, thereby shorting queue times for others. So we can solve two problems at once.
- Installing taps in homes will stretch our resources too thin, so for now, if the queue times are low, we won't improve that source.
2. Most water sources are in rural areas. We need to ensure our teams know this as this means they will have to make these repairs/upgrades in rural areas where road conditions, supplies, and labour are harder challenges to overcome.